runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1423,6 +1535,122 @@ J9::ValuePropagation::doDelayedTransformations()
+   ListIterator<ValueTypesHelperCallTransform> valueTypesHelperCallsToBeFolded(&_valueTypesHelperCallsToBeFolded);
+
+   for (ValueTypesHelperCallTransform *callToTransform = valueTypesHelperCallsToBeFolded.getFirst();
+        callToTransform != NULL;
+        callToTransform = valueTypesHelperCallsToBeFolded.getNext())
+      {
+      TR::TreeTop *callTree = callToTransform->_tree;
+      TR::Node *callNode = callToTransform->_callNode;
+
+      const bool isLoad = callToTransform->_flags.testAny(ValueTypesHelperCallTransform::IsArrayLoad);
+      const bool isStore = callToTransform->_flags.testAny(ValueTypesHelperCallTransform::IsArrayStore);
+      const bool isCompare = callToTransform->_flags.testAny(ValueTypesHelperCallTransform::IsRefCompare);
+      const bool needsStoreCheck = callToTransform->_flags.testAny(ValueTypesHelperCallTransform::RequiresStoreCheck);
+
+      // performTransformation was already checked for comparison non-helper call
+      // Only need to check for array element load or store helper calls
+      if (!isCompare && !performTransformation(
+                            comp(),
+                            "%s Replacing n%dn from acall of <jit%sFlattenableArrayElement> to aloadi\n",
+                            OPT_DETAILS,
+                            callNode->getGlobalIndex(),
+                            isLoad ? "Load" : "Store"))
+         {
+         continue;
+         }
+
+      // Insert dynamic debug counter to describe successful transformation of value type helper or non-helper call
+      if (callToTransform->_flags.testAny(ValueTypesHelperCallTransform::InsertDebugCounter))
+         {
+         const char *operationName = isLoad ? "aaload" : (isStore ? "aastore" : "acmp");
+
+         const char *counterName = TR::DebugCounter::debugCounterName(comp(), "vt-helper/vp-xformed/%s/(%s)/bc=%d",
+                                                               operationName, comp()->signature(), callNode->getByteCodeIndex());
+         TR::DebugCounter::prependDebugCounter(comp(), counterName, callTree);
+         }
+
+      // Transformation for comparison was already handled.  Just needed post-processing to be able to insert debug counter
+      if (isCompare)
+         {
+         continue;
+         }
+
+      TR_ASSERT_FATAL_WITH_NODE(callNode, !comp()->requiresSpineChecks(), "Cannot handle VP yet for jit{Load|Store}FlattenableArrayElement if SpineCHKs are required\n");
+
+      int opIndex = 0;
+
+      TR::Node *valueNode = isLoad ? NULL : callNode->getChild(opIndex++);
+      TR::Node *indexNode = callNode->getChild(opIndex++);
+      TR::Node *arrayRefNode = callNode->getChild(opIndex);
+
+      TR::Node *elementAddressNode = J9::TransformUtil::calculateElementAddress(comp(), arrayRefNode, indexNode, TR::Address);
+
+      if (valueNode != NULL)
+         {
+         valueNode->decReferenceCount();
+         }
+
+      indexNode->decReferenceCount();
+      arrayRefNode->decReferenceCount();
```

Comment:
Thanks for the question, Annabelle @a7ehuo.  It's taken me a while to come back and figure out what this was doing, which is another sign that I should have had some comments here.

The call to `J9::TransformUtil::calculateElementAddress` ends up adding new references to the `indexNode` and `arrayRefNode`.  The call to `recreateWithoutProperties` below recreates the call node with either an `aloadi` or an `awrtbari`, but it won't decrement the reference counts to the original arguments to the call node -- `indexNode`, `arrayRefNode`, and possibly, `valueNode` -- so this code is decrementing the reference counts in preparation for that.

However, it would probably make sense to move the calls to decrement after the call `recreateWithoutProperties`, to make the association more clear.  Also, I see now that the call for `valueNode->decReferenceCount()` might end up making its reference count zero prematurely.

Reply:
Two general rules of thumb around reference counts : 

1. Probably safer to use `recursivelyDecReferenceCount` over just `decReferenceCount` unless you know something about the context to say that a recursive decrement is not needed.
2. If there are reference count operations involving the same nodes wrt calling `recursivelyDecReferenceCount` and `incReferenceCount` then it is preferable to call the `incReferenceCount` operations first to avoid prematurely getting some node's reference count to 0 and then decrementing its children's reference counts as a result. Remember that `incReferenceCount` is not recursive.

Reply:
In commit cfd28b6087da2a2f87c89c904c1b7d5bc57c6567, moved the decrements further down to ensure the reference counts don't prematurely become zero.  Also switched to using `recursivelyDecReferenceCount`.